### PR TITLE
Geocoding

### PR DESCRIPTION
**Pull request by aalices**
This PR adds **geocoding** (address to location) and **reverse geocoding** (location to address) functionalities to Location Module.

Geocoding return a location object that consists of 4 fields latitude, longitude, altitude and accuracy, reverse geocoding's result is an address with following fields if available: city, street region, country, postalCode, name - that is a name of the feature, eg. Big Ben

(comment edited during shipit to remove backticks)

### CODE REVIEW

**Review by terribleben — COMMENTED** [review]
The code looks good on iOS, but I was reading the docs here: https://developer.apple.com/documentation/corelocation/clgeocoder?language=objc and I was wondering about some of the resource considerations mentioned.

- Do we want to auto-cancel pending requests if the experience goes into the background?
- Do we want to offer the developer a method to manually cancel pending requests?
- Do we want to add any more specific error handling around the case where the app is rate-limited?

**Review by jesseruder — COMMENTED** [review]
a few questions on the android side but mostly looks good. thanks!

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: Nope, it doesn't. It uses only the Android one and if the Geocoder is not present an error is thrown with a message `Please check if Geocoder.isPresent() before invoking the search`

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: hmm, it basically removes a listener and I am not sure how to keep track if geocoding is running or not. But I surely have to move this above checking the location parameters

**Comment by aalices** [comment]
@terribleben 
Considering your points:
- The other option would be 'hanging' the coder, so a request will be made after app returns to foreground. Anyway, on Android I stop the coder so maybe we should keep it like that?
- Why not, one can take care of frequency and amount of sent requests.
- If think we can provide some other message when the rate error appears. Should we do anything more?

What's more we also need a check if geocoding is in progress after starting another one as they recommend. A request queue would be solution I guess.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java
Comment (jesseruder) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: ok, it's fine as long as it doesn't error if you call `.stop` when it hasn't been started

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java
Comment (jesseruder) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: ok 👍

**Comment by terribleben** [comment]
- I think stopping the coder in the background to be consistent with Android seems good. Do you know how to hook into foreground/background events in Expo iOS? `EXAccelerometer.m` has a good/simple example.
- Proving a custom rejection error message for rate limiting seems good. 👍 

If a request queue seems good to you, I'm fine with that. If we want to keep it simple for now, that's also fine with me. I think having a good test example would be really useful here... maybe you can add a very simple testing screen for geocoding in our [native-component-list](https://github.com/expo/native-component-list) app? We often use that for testing.

**Comment by jesseruder** [comment]
> I am wondering if it's gonna work like this on Android since it just removes a listener, but the service is running on an intent

can you explain this more? i don't understand

**Comment by arrygoo** [comment]
Excited for this!

**Comment by Cryrivers** [comment]
yep cant wait for it to be merged

**Comment by aalices** [comment]
@jesseruder what I meant was that on iOS you can cancel a request so I suppose it frees some resources and on Android it is not like that so I am wondering if it makes sense to have that able for users. Not sure if my point is right

**Comment by jesseruder** [comment]
ah ok. i don't think we need that before we merge but maybe you can do that in a follow up PR @aalices?

**Comment by giautm** [comment]
Wait for a release with this feature. 👍

**Comment by aalices** [comment]
@jesseruder sure thing

**Review by terribleben — APPROVED** [review]
Thanks! 👍 merging today and we can discuss android followup later

**Comment by terribleben** [comment]
Merged this internally, PR will auto-close when that goes to master.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java
Comment (wli) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: @aalices @jesseruder Unfortunately this seems to break in my environment. Recommend wrapping this line with `Geocoder.isPresent()` or some equivalent?

image

Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/LocationModule.java]: @wli right, I did not considered that particular case just here, will fix it soon, thanks!

**Pull request closed without merging**
